### PR TITLE
Fix Utf8JsonReader.CopyString snippet and split up article

### LINKED ISSUE #35019: Code snippet in `Consume decoded JSON strings` in `How to use a JSON document, Utf8JsonReader, and Utf8JsonWriter in System.Text.Json` does not compile
I copy-pasted [the code snippet from `Consume decoded JSON strings` section](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/use-dom-utf8jsonreader-utf8jsonwriter?pivots=dotnet-7-0#consume-decoded-json-strings) into my visual studio:

``` csharp
int valueLength = reader.HasReadOnlySequence 
    ? checked((int)ValueSequence.Length) 
    : ValueSpan.Length;

char[] buffer = ArrayPool<char>.Shared.Rent(valueLength);
int charsRead = reader.CopyString(buffer);
ReadOnlySpan<char> source = buffer.Slice(0, charsRead);

// Handle the unescaped JSON string.
ParseUnescapedString(source);
ArrayPool<char>.Shared.Return(buffer, clearArray: true);
```

and here is what I got:

image

all the symbols marked in red are not recognized. I searched a bit and looks like this is not a problem with my setup.

For example [Utf8JsonReader](https://learn.microsoft.com/en-us/dotnet/api/system.text.json.utf8jsonreader?view=net-7.0) doesn't appear to have `HasReadOnlySequence` method.

It is also a bit confusing that the section says:

> The following code snippet shows an example of consuming a UTF-16 string using [CopyString](https://learn.microsoft.com/en-us/dotnet/api/system.text.json.utf8jsonreader.copystring).

but refers to UTF-8. So is it UTF-8 or UTF-16 ?

I think the doc should provide near-obviously-compilable code snippets. At the very least it should explain what needs to be done to make the snippet compile, like what dependencies to import to get the right `Slice` method, how to implement `ParseUnescapedString` and so on.

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: 73a13d38-4904-c068-c9eb-e43176ce5338
* Version Independent ID: c5dfa4d2-c77a-e752-dae7-0bfae5756b76
* Content: [How to use a JSON document, Utf8JsonReader, and Utf8JsonWriter in System.Text.Json](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/use-dom-utf8jsonreader-utf8jsonwriter)
* Content Source: [docs/standard/serialization/system-text-json/use-dom-utf8jsonreader-utf8jsonwriter.md](https://github.com/dotnet/docs/blob/main/docs/standard/serialization/system-text-json/use-dom-utf8jsonreader-utf8jsonwriter.md)
* Product: **dotnet-fundamentals**
* GitHub Login: @gewarren
* Microsoft Alias: **gewarren**

### PR DESCRIPTION
**Pull request by gewarren**
Fixes #35019.

- Fixes CopyString snippet and puts it into separate code file so it can be compiled.
- Splits the Json DOM, Utf8JsonReader, and Utf8JsonWriter article into three separate articles and updates backlinks as necessary.

image

- Removes large See Also lists (they are hard to maintain, likely too long for most people to read, and people can use the TOC to navigate).

---

#### Internal previews

<details><summary><strong>Toggle expand/collapse</strong></summary><br/>

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/core/whats-new/dotnet-6.md](https://github.com/dotnet/docs/blob/97b90c8f8746eb27d4ffe28bde5e6a3162997ac9/docs/core/whats-new/dotnet-6.md) | [What's new in .NET 6](https://review.learn.microsoft.com/en-us/dotnet/core/whats-new/dotnet-6?branch=pr-en-us-35489) |
| [docs/standard/serialization/system-text-json/character-casing.md](https://github.com/dotnet/docs/blob/97b90c8f8746eb27d4ffe28bde5e6a3162997ac9/docs/standard/serialization/system-text-json/character-casing.md) | [How to enable case-insensitive property name matching with System.Text.Json](https://review.learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/character-casing?branch=pr-en-us-35489) |
| [docs/standard/serialization/system-text-json/character-encoding.md](https://github.com/dotnet/docs/blob/97b90c8f8746eb27d4ffe28bde5e6a3162997ac9/docs/standard/serialization/system-text-json/character-encoding.md) | [How to customize character encoding with System.Text.Json](https://review.learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/character-encoding?branch=pr-en-us-35489) |
| [docs/standard/serialization/system-text-json/converters-how-to.md](https://github.com/dotnet/docs/blob/97b90c8f8746eb27d4ffe28bde5e6a3162997ac9/docs/standard/serialization/system-text-json/converters-how-to.md) | [How to write custom converters for JSON serialization (marshalling) in .NET](https://review.learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/converters-how-to?branch=pr-en-us-35489) |
| [docs/standard/serialization/system-text-json/customize-properties.md](https://github.com/dotnet/docs/blob/97b90c8f8746eb27d4ffe28bde5e6a3162997ac9/docs/standard/serialization/system-text-json/customize-properties.md) | [docs/standard/serialization/system-text-json/customize-properties](https://review.learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/customize-properties?branch=pr-en-us-35489) |
| [docs/standard/serialization/system-text-json/handle-overflow.md](https://github.com/dotnet/docs/blob/97b90c8f8746eb27d4ffe28bde5e6a3162997ac9/docs/standard/serialization/system-text-json/handle-overflow.md) | [How to handle overflow JSON or use JsonElement or JsonNode in System.Text.Json](https://review.learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/handle-overflow?branch=pr-en-us-35489) |
| [docs/standard/serialization/system-text-json/how-to.md](https://github.com/dotnet/docs/blob/97b90c8f8746eb27d4ffe28bde5e6a3162997ac9/docs/standard/serialization/system-text-json/how-to.md) | ["How to serialize and deserialize JSON using C# - .NET"](https://review.learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/how-to?branch=pr-en-us-35489) |
| [docs/standard/serialization/system-text-json/ignore-properties.md](https://github.com/dotnet/docs/blob/97b90c8f8746eb27d4ffe28bde5e6a3162997ac9/docs/standard/serialization/system-text-json/ignore-properties.md) | [How to ignore properties with System.Text.Json](https://review.learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/ignore-properties?branch=pr-en-us-35489) |
| [docs/standard/serialization/system-text-json/immutability.md](https://github.com/dotnet/docs/blob/97b90c8f8746eb27d4ffe28bde5e6a3162997ac9/docs/standard/serialization/system-text-json/immutability.md) | [docs/standard/serialization/system-text-json/immutability](https://review.learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/immutability?branch=pr-en-us-35489) |
| [docs/standard/serialization/system-text-json/invalid-json.md](https://github.com/dotnet/docs/blob/97b90c8f8746eb27d4ffe28bde5e6a3162997ac9/docs/standard/serialization/system-text-json/invalid-json.md) | [How to allow some kinds of invalid JSON with System.Text.Json](https://review.learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/invalid-json?branch=pr-en-us-35489) |
| [docs/standard/serialization/system-text-json/migrate-from-newtonsoft.md](https://github.com/dotnet/docs/blob/97b90c8f8746eb27d4ffe28bde5e6a3162997ac9/docs/standard/serialization/system-text-json/migrate-from-newtonsoft.md) | ["Migrate from Newtonsoft.Json to System.Text.Json - .NET"](https://review.learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/migrate-from-newtonsoft?branch=pr-en-us-35489) |
| [docs/standard/serialization/system-text-json/overview.md](https://github.com/dotnet/docs/blob/97b90c8f8746eb27d4ffe28bde5e6a3162997ac9/docs/standard/serialization/system-text-json/overview.md) | [JSON serialization and deserialization (marshalling and unmarshalling) in .NET - overview](https://review.learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/overview?branch=pr-en-us-35489) |
| [docs/standard/serialization/system-text-json/polymorphism.md](https://github.com/dotnet/docs/blob/97b90c8f8746eb27d4ffe28bde5e6a3162997ac9/docs/standard/serialization/system-text-json/polymorphism.md) | [docs/standard/serialization/system-text-json/polymorphism](https://review.learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/polymorphism?branch=pr-en-us-35489) |
| [docs/standard/serialization/system-text-json/preserve-references.md](https://github.com/dotnet/docs/blob/97b90c8f8746eb27d4ffe28bde5e6a3162997ac9/docs/standard/serialization/system-text-json/preserve-references.md) | [How to preserve references and handle or ignore circular references in System.Text.Json](https://review.learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/preserve-references?branch=pr-en-us-35489) |
| [docs/standard/serialization/system-text-json/source-generation-modes.md](https://github.com/dotnet/docs/blob/97b90c8f8746eb27d4ffe28bde5e6a3162997ac9/docs/standard/serialization/system-text-json/source-generation-modes.md) | [How to choose reflection or source generation in System.Text.Json](https://review.learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/source-generation-modes?branch=pr-en-us-35489) |
| [docs/standard/serialization/system-text-json/source-generation.md](https://github.com/dotnet/docs/blob/97b90c8f8746eb27d4ffe28bde5e6a3162997ac9/docs/standard/serialization/system-text-json/source-generation.md) | [How to use source generation in System.Text.Json](https://review.learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/source-generation?branch=pr-en-us-35489) |
| [docs/standard/serialization/system-text-json/supported-collection-types.md](https://github.com/dotnet/docs/blob/97b90c8f8746eb27d4ffe28bde5e6a3162997ac9/docs/standard/serialization/system-text-json/supported-collection-types.md) | [docs/standard/serialization/system-text-json/supported-collection-types](https://review.learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/supported-collection-types?branch=pr-en-us-35489) |
| [docs/standard/serialization/system-text-json/use-dom.md](https://github.com/dotnet/docs/blob/97b90c8f8746eb27d4ffe28bde5e6a3162997ac9/docs/standard/serialization/system-text-json/use-dom.md) | [docs/standard/serialization/system-text-json/use-dom](https://review.learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/use-dom?branch=pr-en-us-35489) |
| [docs/standard/serialization/system-text-json/use-utf8jsonreader.md](https://github.com/dotnet/docs/blob/97b90c8f8746eb27d4ffe28bde5e6a3162997ac9/docs/standard/serialization/system-text-json/use-utf8jsonreader.md) | [docs/standard/serialization/system-text-json/use-utf8jsonreader](https://review.learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/use-utf8jsonreader?branch=pr-en-us-35489) |
| [docs/standard/serialization/system-text-json/use-utf8jsonwriter.md](https://github.com/dotnet/docs/blob/97b90c8f8746eb27d4ffe28bde5e6a3162997ac9/docs/standard/serialization/system-text-json/use-utf8jsonwriter.md) | [docs/standard/serialization/system-text-json/use-utf8jsonwriter](https://review.learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/use-utf8jsonwriter?branch=pr-en-us-35489) |
| [docs/whats-new/dotnet-7-docs.md](https://github.com/dotnet/docs/blob/97b90c8f8746eb27d4ffe28bde5e6a3162997ac9/docs/whats-new/dotnet-7-docs.md) | [.NET docs: What's new for the .NET 7 release](https://review.learn.microsoft.com/en-us/dotnet/whats-new/dotnet-7-docs?branch=pr-en-us-35489) |

</details>

### CODE REVIEW

File: docs/standard/serialization/system-text-json/snippets/system-text-json-how-to/csharp/Utf8ReaderCopyString.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+            // <Snippet1>
+            var reader = new Utf8JsonReader( /* jsonReadOnlySpan */ );
+
+            int valueLength = reader.HasValueSequence
+                ? checked((int)reader.ValueSequence.Length)
+                : reader.ValueSpan.Length;
+
+            char[] buffer = ArrayPool<char>.Shared.Rent(valueLength);
+            int charsRead = reader.CopyString(buffer);
+            ReadOnlySpan<char> source = buffer.AsSpan(0, charsRead);
```
Comment (gewarren) [inline docs/standard/serialization/system-text-json/snippets/system-text-json-how-to/csharp/Utf8ReaderCopyString.cs:19]: @eiriktsarpalis I changed this from buffer.Slice to buffer.AsSpan. Is that correct?

Comment (eiriktsarpalis) [reply docs/standard/serialization/system-text-json/snippets/system-text-json-how-to/csharp/Utf8ReaderCopyString.cs:19]: Yes, I think the previous snippet wouldn't compile actually.

**Review by eiriktsarpalis — APPROVED** [review]
Thanks!

**Review by BillWagner — APPROVED** [review]
This LGTM @gewarren 

Let's :shipit:

**Pull request merged**